wgpu-hal/src/dx12/shader_compilation.rs

```diff
@@ -75,234 +73,237 @@ pub(super) fn compile_fxc(
+        .into_device_result("GetOutput")?;
+    result__.ok_or(crate::DeviceError::Unexpected)
+}
+
+fn as_err_str(blob: &Dxc::IDxcBlobUtf8) -> Result<&str, crate::DeviceError> {
+    let ptr = unsafe { blob.GetStringPointer() };
+    let len = unsafe { blob.GetStringLength() };
+    core::str::from_utf8(unsafe { core::slice::from_raw_parts(ptr.0, len) })
+        .map_err(|_| crate::DeviceError::Unexpected)
+}
+
+pub(super) fn compile_dxc(
+    device: &crate::dx12::Device,
+    source: &str,
+    source_name: Option<&CStr>,
+    raw_ep: &str,
+    stage_bit: wgt::ShaderStages,
+    full_stage: &str,
+    dxc_container: &DxcContainer,
+) -> Result<crate::dx12::CompiledShader, crate::PipelineError> {
+    profiling::scope!("compile_dxc");
+
+    let source_name = source_name.and_then(|cstr| cstr.to_str().ok());
+
+    let source_name = source_name.map(OPCWSTR::new);
+    let raw_ep = OPCWSTR::new(raw_ep);
+    let full_stage = OPCWSTR::new(full_stage);
+
+    let mut compile_args = arrayvec::ArrayVec::<PCWSTR, 12>::new_const();
+
+    if let Some(source_name) = source_name.as_ref() {
+        compile_args.push(source_name.ptr())
     }
 
-    // It shouldn't be possible that this gets called with the `dxc_shader_compiler` feature disabled.
-    pub(crate) fn compile_dxc(
-        _device: &crate::dx12::Device,
-        _source: &str,
-        _source_name: Option<&CStr>,
-        _raw_ep: &str,
-        _stage_bit: wgt::ShaderStages,
-        _full_stage: String,
-        _dxc_container: &DxcContainer,
-    ) -> (
-        Result<crate::dx12::CompiledShader, crate::PipelineError>,
-        log::Level,
-    ) {
-        unimplemented!("Something went really wrong, please report this. Attempted to compile shader with DXC, but the DXC feature is disabled. Enable the `dxc_shader_compiler` feature on wgpu_hal to use DXC.");
+    compile_args.extend([
+        windows::core::w!("-E"),
+        raw_ep.ptr(),
+        windows::core::w!("-T"),
+        full_stage.ptr(),
+        windows::core::w!("-HV"),
+        windows::core::w!("2018"), // Use HLSL 2018, Naga doesn't supported 2021 yet.
+        windows::core::w!("-no-warnings"),
+        Dxc::DXC_ARG_ENABLE_STRICTNESS,
+        Dxc::DXC_ARG_SKIP_VALIDATION, // Disable implicit validation to work around bugs when dxil.dll isn't in the local directory.
+    ]);
+
+    if device
+        .private_caps
+        .instance_flags
+        .contains(wgt::InstanceFlags::DEBUG)
+    {
+        compile_args.push(Dxc::DXC_ARG_DEBUG);
+        compile_args.push(Dxc::DXC_ARG_SKIP_OPTIMIZATIONS);
     }
+
+    let buffer = Dxc::DxcBuffer {
+        Ptr: source.as_ptr().cast(),
+        Size: source.len(),
+        Encoding: Dxc::DXC_CP_UTF8.0,
+    };
+
+    let compile_res: Dxc::IDxcResult = unsafe {
+        dxc_container
+            .compiler
+            .Compile(&buffer, Some(&compile_args), None)
+    }
+    .into_device_result("Compile")?;
```

Comment:
Like the old API, I wonder if `Compile` returns an error code while `IDxcResult` is _also_ returned with valid errors (the root cause) in `DXC_OUT_ERRORS`.

Or if that's still communicated by `IDxcOperationResult::GetResult()` which you're not checking here?

Reply:
From my testing, `IDxcCompiler3::Compile` seems to succeed even if there was a compilation error.

Getting the blob via `IDxcResult::GetOutput(DXC_OUT_ERRORS)` always succeeds. If there is a compilation error its size/length will be non 0 and will contain the error message.

The wiki seems to say that `Compile` will only fail if the compiler itself ran into an internal error.
https://github.com/microsoft/DirectXShaderCompiler/wiki/Advanced-Error-Handling#internal-compiler-errors